MAINT: Converge on one shared Font class for text extraction and appearance streams

PJBrs opened the pull request:
EDITED ON 2 JANUARY

This PR introduces one dedicated Font class for all text extraction code and appearance streams. This should be a basis for adding new font resources to appearance streams, in relation to bug https://github.com/py-pdf/pypdf/issues/3514

Changes:
- Adds new font class to replace the one in the text extraction code
- Uses resulting font class in the layout mode text extraction code page.py
- Uses resulting font class in the Appearance Stream code
- Includes character widths for string encodings, which were previously omitted
- Adds default width to character widths
- Uses the resulting font class in the non-layout mode text extraction code in place of cmap
- Removes obsolete code from cmap.py

This PR got rather big, because, first, I noticed how the original font class did not parse widths for some string encoded fonts, then I noticed that it didn't set a default width either. However, this resulted in rather low coverage. To resolve low coverage I also ported the non-layout mode text extraction code to the new Font class. This was a lot more work, but now coverage is very good, without the need to add more tests! The new font class works fine with the current appearance stream code and both the original and the layout mode text extraction code.

The current version - January 2 - is ready for review. I'm specifically looking for feedback on:
- Is this PR too big? If so, I can logically limit it to, for instance, the first seven patches. This excludes replacing the cmaps in text extraction, and therefore results in lower test coverage (for the time being).

Small note for reviewing - the diff stat for this PR is rather big, but the relevant changes are mostly limited to _font.py and the _get_actual_text_widths method in ‎pypdf/_text_extraction/_text_extractor.py (https://github.com/py-pdf/pypdf/pull/3583/changes/012826d0a11932de8a1622b35c1ad3c3ab928e89). The rest is only restructuring, mostly renaming and re-typing variables.

Some further advantages of this PR:
1. This PR _adds_ functionality to the Font class for the layout mode text extraction code: parsing character widths for Type1 and TrueType fonts with string encodings
2. This patch removes code duplication in the sense that build_font_width_map did more or less the same as the original Font class
3. Increases robustness because all text extraction code (original and layout mode) _and_ the appearance stream code use the same basis for font information
4. Net removal of more than 150 lines of code (see below)
5. Text extraction is faster with the new code (see further below)
6. Improves readability, because the original cmap tuple information can now be accessed as font.encoding, font.character_map, font.space_width and font_resource, instead of cmap[0], cmap[1], cmap[2] and cmap[3].

```
$ git diff --stat origin/main
 pypdf/_cmap.py                                             | 198 --------------------------------------------------------------------------------------------------------
 pypdf/_font.py                                             | 225 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------
 pypdf/_page.py                                             |  52 ++++++++++++----------------
 pypdf/_text_extraction/__init__.py                         |  30 +++++++---------
 pypdf/_text_extraction/_layout_mode/__init__.py            |   2 +-
 pypdf/_text_extraction/_layout_mode/_fixed_width_page.py   |   2 +-
 pypdf/_text_extraction/_layout_mode/_font.py               |  65 -----------------------------------
 pypdf/_text_extraction/_layout_mode/_text_state_manager.py |   4 +--
 pypdf/_text_extraction/_layout_mode/_text_state_params.py  |  10 +++---
 pypdf/_text_extraction/_text_extractor.py                  | 144 ++++++++++++++++++++++++++++++++++------------------------------------------
 pypdf/generic/_appearance_stream.py                        |  63 +++++++++++++++++-----------------
 resources/010-pdflatex-forms.txt                           |   2 +-
 resources/multicolumn-lorem-ipsum.txt                      |  80 +++++++++++++++++++++---------------------
 tests/test_cmap.py                                         |   8 +++--
 tests/test_text_extraction.py                              |  33 +++++++++---------
 15 files changed, 382 insertions(+), 536 deletions(-)
```

```
time (for i in $(seq 10); do pytest3 tests/test_text_extraction.py ; done)
Old:
real    0m56.213s
user    0m45.754s
sys     0m10.463s

New:
real    0m52.308s
user    0m42.508s
sys     0m9.802s
```

stefan6419846 commented on pypdf/_text_extraction/_layout_mode/_text_state_params.py:
```diff
@@ -72,9 +72,9 @@ def __post_init__(self) -> None:
-            self.space_tx = round(self.word_tx("", self.font.space_width * -2), 3)
+            # a TD_offset of -1 * font.space_width where font.space_width is
+            # the space_width calculated in _font.py.
+            self.space_tx = round(self.word_tx("", -self.font.space_width), 3)
```
Why are you using a new factor here?

stefan6419846 commented on pypdf/_text_extraction/__init__.py:
```diff
@@ -233,14 +229,14 @@ def get_display_str(
                 rtl_dir = False
                 if visitor_text is not None:
-                    visitor_text(text, cm_matrix, tm_matrix, cmap[3], font_size)
+                    visitor_text(text, cm_matrix, tm_matrix, font_resource, font_size)
```
Similar to https://github.com/py-pdf/pypdf/pull/3583/changes#r2661249710

PJBrs replied:
See above.

stefan6419846 commented on pypdf/_text_extraction/_text_extractor.py:
```diff
@@ -85,15 +86,17 @@ def __init__(self) -> None:
-        )  # (encoding, CMAP, font resource name, font)
+        self.font_resource: Optional[DictionaryObject] = None
+        self.font = Font(
+                name = "NotInitialized",
```
These lines are indented a level too much.

PJBrs replied:
Done.

stefan6419846 commented on pypdf/_text_extraction/_text_extractor.py:
```diff
@@ -217,22 +203,30 @@ def _handle_tj(
-            self._get_actual_font_widths(cmap, text_operands, font_size, space_width))
+            self._get_actual_text_widths(
+                font,
+                text.replace(previous_text, ""),
```
Could you please elaborate on this?

stefan6419846 commented on pypdf/generic/_appearance_stream.py:
```diff
@@ -183,11 +184,11 @@ def _generate_appearance_stream_data(
 
         """
         rectangle = self._layout.rectangle
+        font = cast(Font, font)
```
This looks wrong. What happens for `font is None`?

PJBrs replied:
Font is never None here. If we do have a font resource before calling _generate_appearance_stream_data, then we use font = Font.from_font_resource(font_resource) to get the Font. This method does not return None.

If we don't have a Font resource, then we create this default:

            font = Font(
                name="Helvetica",
                character_map={},
                encoding=dict(zip(range(256), fill_from_encoding("cp1252"))),  # WinAnsiEncoding
                sub_type="Type1",
                font_descriptor = font_descriptor,
                character_widths = font_descriptor.character_widths
            )
Either way, we have initialised a Font once we get here. The cast was to silence mypy. It was there before as well, as:
font_descriptor = cast(FontDescriptor, font_descriptor)

I can see though that it is somewhat strange that the _generate_appearance_stream_data types the font as Optional[Font].. ... I remember though; the alternative would be to set a default value for Font, and we wanted to prevent that as well.

In sum: we decided to type font as Optional so we could default as None, then we needed to cast font as Font to silence mypy. But it never was optional anyway.

PJBrs replied:
Done.

stefan6419846 commented on pypdf/_font.py:
```diff
@@ -136,58 +183,146 @@ def from_font_resource(
+                cls._add_default_width(font_descriptor.character_widths)
+                return font_descriptor
+
+            if "/FontDescriptor" in pdf_font_dict:  # TODO: This does not account for some Type3 fonts;
```
Is there something to do about this?

PJBrs replied:
Yes, quite easily. I didn't want to complicate this PR any further,  but I can definitely add code for this.

PJBrs commented on pypdf/_text_extraction/__init__.py:
```diff
@@ -137,7 +136,7 @@ def crlf_space_check(
                         memo_cm,
                         memo_tm,
-                        cmap[3],
+                        font_resource,
```
It shouldn't, because it is the exact same object.

PJBrs commented on pypdf/_text_extraction/__init__.py:
```diff
@@ -233,14 +229,14 @@ def get_display_str(
                 rtl_dir = True
                 if visitor_text is not None:
-                    visitor_text(text, cm_matrix, tm_matrix, cmap[3], font_size)
+                    visitor_text(text, cm_matrix, tm_matrix, font_resource, font_size)
```
See above.

stefan6419846 commented on pypdf/generic/_appearance_stream.py:
```diff
@@ -78,7 +78,7 @@ def _scale_text(
 
         Args:
-            font_descriptor: A FontDescriptor for the font to be used.
+            font: The Font to be used.
```
I guess this went to the wrong comment? Nevertheless, the correct way would be to update the method signature in this case and ensure that all callers are passing the correct values, instead of this wrong cast.

PJBrs replied:
Yeah sorry. See below for explanation.

PJBrs replied:
Done.

PJBrs commented on pypdf/_font.py:
```diff
@@ -136,58 +183,146 @@ def from_font_resource(
+        ):
+            d_font = cast(DictionaryObject, d_font.get_object())
+            cast(ArrayObject, pdf_font_dict["/DescendantFonts"])[d_font_idx] = d_font
+            # Collect character widths
```
Yes, I removed all the comments about collecting character widths and  font descriptors.

PJBrs commented on pypdf/_text_extraction/_text_extractor.py:
```diff
@@ -317,39 +313,26 @@ def _handle_tf(self, operands: list[Any]) -> None:
+            font_descriptor = FontDescriptor()
+            self.font = Font(
+                "Unknown",
+                space_width=19998,
```
https://github.com/py-pdf/pypdf/blob/main/pypdf/_cmap.py#L76C1-L82C2

This if the Font translation of the unknown_char_map tuple in _cmap.py. unknown_char_map[1] is the space_width, but in terms of build_char_map, this represents half a space, and 19998 = 2 * 9999.

I'll admit that it looks rather weird, I could have just chosen 250.

PJBrs replied:
Changed to 250.

PJBrs commented on pypdf/_font.py:
```diff
@@ -136,58 +183,146 @@ def from_font_resource(
+                return font_descriptor
+
+            if "/FontDescriptor" in pdf_font_dict:  # TODO: This does not account for some Type3 fonts;
+                                                    # see tests/test_cmap.py::test_ascii_charset
```
Changed for now though.

PJBrs commented on pypdf/_font.py:
```diff
@@ -136,58 +183,146 @@ def from_font_resource(
+    @staticmethod
+    def to_dict(font_instance: "Font") -> dict[str, Any]:
+        """Dataclass to dict for json.dumps serialization."""
+        return {
```
This is the method where it is used:
```
       if debug_path:  # pragma: no cover
            import json  # noqa: PLC0415

            debug_path.joinpath("fonts.json").write_text(
                json.dumps(
                    fonts, indent=2, default=lambda x: getattr(x, "to_dict", str)(x)
                ),
                "utf-8",
            )
```

I changed the important part to:
```
            debug_path.joinpath("fonts.json").write_text(
                json.dumps(fonts, indent=2, default=asdict),
                "utf-8"
            )
```

I DID NOT TEST THE RESULT. I did change the associated test though to use asdict as well.

PJBrs replied:
Tested, works.

PJBrs commented on pypdf/_font.py:
```diff
@@ -136,58 +183,146 @@ def from_font_resource(
+
+    name: str = "Unknown"
+    character_map: dict[Any, Any] = field(default_factory=dict)
+    encoding: Union[str, dict[int, str]] = "utf-16-be"
```
And only now I understand your point - why set a default at all. Removed.

PJBrs commented:
@stefan6419846 Thanks for your review! I addressed most of your points.

I did make a couple of additional changes. In the Font class, I changed this:
```
@@ -317,12 +322,5 @@ class Font:
     def text_width(self, text: str = "") -> float:
         """Sum of character widths specified in PDF font for the supplied text."""
         return sum(
-            [self.character_widths.get(char, self.space_width) for char in text], 0.0
+            [self.character_widths.get(char, self.character_widths["default"]) for char in text], 0.0
         )
```

The underlying logic is as follows - when a default character width is missing, then you should fall back to default width, not space width. As a rule, assume that default width is roughly the width of two spaces.

This causes problems in the layout mode font class, which, as I mentioned earlier, assumed that what was passed as half a space's width actually represented a full space width. That's why the original (wrong) code worked, while the new caused a bit of trouble. So I added the following, which also exists in the non-layout mode code:
```
@@ -117,8 +117,14 @@ class TextStateParams:
 
     def word_tx(self, word: str, TD_offset: float = 0.0) -> float:
         """Horizontal text displacement for any word according this text state"""
+        width: float = 0.0
+        for char in word:
+            if char == " ":
+                width += self.font.space_width
+            else:
+                width += self.font.text_width(char)
         return (
-            (self.font_size * ((self.font.text_width(word) - TD_offset) / 1000.0))
+            (self.font_size * ((width - TD_offset) / 1000.0))
             + self.Tc
             + word.count(" ") * self.Tw
         ) * (self.Tz / 100.0)
```
This fixes all tests, except one, where I think the changed rendering is actually correct. See this commit's message for explanation: https://github.com/py-pdf/pypdf/pull/3583/commits/89cc31057543a23893aeedb61dde98e9f04cb319. In short, there is one important way in which the new layout mode code deviates from the old code, and that's the fallback width in the Font text_width code. This changes rendering in case we encounter a character with unknown width that is not a space. 

Finally, in the non-layout extraction code, I noticed two things:
 - I was still passing around space_width while this was already part of the font class.
 - I did not take into account user-provided space width anymore. 

I removed all that passing space width around, and added the following in _page.py, to restore the old behaviour:
```
@@ -1715,6 +1715,9 @@ class PageObject(DictionaryObject):
                     font_resource_object = cast(DictionaryObject, font_resources_dict[font_resource].get_object())
                     font_resources[font_resource] = font_resource_object
                     fonts[font_resource] = Font.from_font_resource(font_resource_object)
+                    # Override space width, if applicable
+                    if fonts[font_resource].character_widths.get(" ", 0) == 0:
+                        fonts[font_resource].space_width = space_width
                 except (AttributeError, TypeError):
                     pass
 
```

stefan6419846 commented on pypdf/_text_extraction/_text_extractor.py:
```diff
@@ -164,84 +165,69 @@ def _post_process_text_operation(self, str_widths: float) -> None:
-                font_widths += compute_font_width(font_width_map, char)
-        return (font_widths * font_size, space_width * font_size, font_size)
+                font_widths += font.text_width(char)
+        return (font_widths * font_size, font_size)
```
```suggestion
        return font_widths * font_size, font_size
```

PJBrs replied:
Here's the new approach: https://github.com/py-pdf/pypdf/pull/3583/commits/b4c1a14b0d00cf92c0ce6c6b735ca1a50cd64427

PJBrs commented on pypdf/generic/_appearance_stream.py:
```diff
@@ -142,9 +142,9 @@ def _scale_text(
-        selection: Optional[list[str]] = None,
-        font_descriptor: Optional[FontDescriptor] = None,
+        text: str,
+        selection: Optional[list[str]],
```
In principle, I agree, but it is only a private method and for readability I'd like to propose to keep it as is, with similar arguments (in this case: text and selection) right after each other. Of course, if you disagree, I can also just change the order of the arguments.

stefan6419846 replied:
Currently, you always have to supply a value. We can keep the order like this, but then please make this `Union[list[str], None]` instead - for typing, this should have the same meaning, but it is more obvious that having either `None` or a list of strings here is the desired behavior, not being able to potentially omit this parameter.

PJBrs commented on pypdf/_text_extraction/_layout_mode/_text_state_params.py:
```diff
@@ -117,8 +117,14 @@ def displacement_matrix(
 
     def word_tx(self, word: str, TD_offset: float = 0.0) -> float:
```
Will do!

stefan6419846 reviewed (approved):
Thanks for your patience.
